### ISSUE TITLE
Invalid header data may lead to unsafe memory reads

### ISSUE THREAD

**Issue by 197g** [ababo/dtb#5]
The current implementation trusts, to some extent, the offsets provided by the
file header to access its buffers. In particular, the order of checks for ensuring
that offsets and sizes are well defined on the whole blob is not enough to 
avoid wrong all invalid indices.

For example `reserved_mem_offset` is used to read from the buffer first but 
the order in `Reader::read` only checks the `total_size` last. Furthermore,
several conversion with `as usize` and additions may lead to overflows that
could confuse the conversion.

The impact for the `struct` and `strings` offsets is a simple panic, as they are 
used as slice indices only which are fully checked against the buffer length. 
However, `reserved_mem_offset` is used to `unsafe`ly construct a `&[u8]` 
before any of the other checks have been performed. This allows an attacker
to create files with an invalid out-of-bounds `reserved_mem_offset` that is
improperly check in that portion of code.

It is then used to manipulate a pointer to the input buffer:
https://github.com/ababo/dtb/blob/3ca29eff4e29df50a45713bf954edcfb493ab5f7/src/reader.rs#L365

An invalid offset may thus lead to this pointer pointing out-of-bounds and 
overflow in the length calculation will then iterate over a large portion of 
invalid memory.

**Comment by 197g** [comment]
Since Github doesn't allow arbitrary file uploads, here's a hexdump of a faulty file that would lead to a segmentation fault when ran through `dump` example:

```
00000000  d0 0d fe ed 00 00 00 28  00 00 00 58 00 00 01 74  |.......(...X...t|
00000010  ff ff ff f8 00 00 00 11  00 00 00 10 00 00 00 00  |................|
00000020  00 00 00 8b 00 00 01 1c                           |........|
00000028
```

**Comment by ababo** [comment]
Fixed in #4
